faunadb-java-dsl/src/main/java/com/faunadb/client/types/Value.java

```diff
@@ -505,31 +503,168 @@ public SetRefV(@JsonProperty("@set") ImmutableMap<String, Value> parameters) {
     }
   }
 
+  @JsonInclude(JsonInclude.Include.NON_ABSENT)
+  public static class RefID {
+    @JsonProperty("id")       private final String id;
+    @JsonProperty("class")    private final RefV clazz;
+    @JsonProperty("database") private final RefV database;
+
+    private RefID(String id, RefV clazz, RefV database) {
+      this.id = id;
+      this.clazz = clazz;
+      this.database = database;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj instanceof RefID))
+        return false;
+
+      RefID other = (RefID) obj;
+
+      return Objects.equal(id, other.id) &&
+        Objects.equal(clazz, other.clazz) &&
+        Objects.equal(database, other.database);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(id, clazz, database);
+    }
+  }
+
   /**
    * A FaunaDB ref type.
    *
    * @see <a href="https://fauna.com/documentation/queries#values-special_types">FaunaDB Special Types</a>
    */
-  public static final class RefV extends Value.ScalarValue<String> {
+  public static final class RefV extends Value.ScalarValue<RefID> {
 
-    @JsonCreator
-    public RefV(@JsonProperty("@ref") String value) {
-      super(value);
+    public RefV(String id, RefV clazz, RefV database) {
+      super(new RefID(id, clazz, database));
+    }
+
+    public RefV(String id, RefV clazz) {
+      this(id, clazz, null);
+    }
+
+    @JsonIgnore
+    public String getId() {
+      return value.id;
+    }
+
+    @JsonIgnore
+    public Optional<RefV> getClazz() {
+      return Optional.fromNullable(value.clazz);
+    }
+
+    @JsonIgnore
+    public Optional<RefV> getDatabase() {
+      return Optional.fromNullable(value.database);
     }
 
     @Override
     @JsonProperty("@ref")
-    protected String toJson() {
+    protected Object toJson() {
       return value;
     }
 
-    /**
-     * Extracts its string value.
-     *
-     * @return a string with the ref value
-     */
-    public String strValue() {
-      return value;
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj instanceof RefV))
+        return false;
+
+      RefV other = (RefV) obj;
+
+      return Objects.equal(value, other.value);
+    }
+
+    @Override
+    public int hashCode() {
+      return value.hashCode();
+    }
+
+    @Override
+    public String toString() {
+      String cls = value.clazz != null ? format(", class = %s", value.clazz) : "";
+      String db = value.database != null ? format(", database = %s", value.database) : "";
+      return format("%s(id = \"%s\"%s%s)", getClass().getSimpleName(), value.id, cls, db);
+    }
+  }
+
+  public static RefV RefV(String id, RefV clazz) {
```

Comment:
I think the driver should guid people towards the expression constructors instead. Having queries that mixes values and expressions can be very confusing. 

I'd leave this value constructors out unless there is a significant advantage that I'm not aware of.